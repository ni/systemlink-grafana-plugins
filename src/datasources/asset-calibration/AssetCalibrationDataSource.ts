import {
  DataFrameDTO,
  DataQueryRequest,
  DataSourceInstanceSettings,
  FieldDTO,
  TestDataSourceResponse,
} from '@grafana/data';
import { BackendSrv, getBackendSrv, getTemplateSrv, TemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import {
  AssetCalibrationForecastKey,
  AssetCalibrationQuery,
  AssetCalibrationTimeBasedGroupByType,
  CalibrationForecastResponse,
  ColumnDescriptorType,
  FieldDTOWithDescriptor,
} from './types';
import { transformComputedFieldsQuery } from 'core/query-builder.utils';
import { AssetComputedDataFields } from './constants';
import { AssetModel, AssetsResponse } from 'datasources/asset-common/types';
import TTLCache from '@isaacs/ttlcache';
import { metadataCacheTTL } from 'datasources/data-frame/constants';
import { SystemMetadata } from 'datasources/system/types';
import { defaultOrderBy, defaultProjection } from 'datasources/system/constants';

export class AssetCalibrationDataSource extends DataSourceBase<AssetCalibrationQuery> {
  public defaultQuery = {
    groupBy: [],
    filter: ''
  };

  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv(),
  ) {
    super(instanceSettings, backendSrv, templateSrv);
  }

  baseUrl = this.instanceSettings.url + '/niapm/v1';

  systemAliasCache: TTLCache<string, SystemMetadata> = new TTLCache<string, SystemMetadata>({ ttl: metadataCacheTTL });

  async runQuery(query: AssetCalibrationQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
    await this.loadSystems();

    if (query.filter) {
      query.filter = this.templateSrv.replace(transformComputedFieldsQuery(query.filter, AssetComputedDataFields), options.scopedVars);
    }

    return await this.processCalibrationForecastQuery(query as AssetCalibrationQuery, options);
  }

  async processCalibrationForecastQuery(query: AssetCalibrationQuery, options: DataQueryRequest) {
    const result: DataFrameDTO = { refId: query.refId, fields: [] };
    const from = options.range!.from.toISOString();
    const to = options.range!.to.toISOString();

    const calibrationForecastResponse: CalibrationForecastResponse = await this.queryCalibrationForecast(query.groupBy, from, to, query.filter);

    result.fields = calibrationForecastResponse.calibrationForecast.columns || [];
    if (this.isGroupByTime(query)) {
      this.processResultsGroupedByTime(result)
    } else {
      this.processResultsGroupedByProperties(result)
    }

    return result;
  }

  isGroupByTime(query: AssetCalibrationQuery) {
    return query.groupBy.includes(AssetCalibrationTimeBasedGroupByType.Day) ||
      query.groupBy.includes(AssetCalibrationTimeBasedGroupByType.Week) ||
      query.groupBy.includes(AssetCalibrationTimeBasedGroupByType.Month);
  }

  processResultsGroupedByTime(result: DataFrameDTO) {
    result.fields.forEach(field => {
      field.name = this.createColumnNameFromDescriptor(field as FieldDTOWithDescriptor);
      switch (field.name) {
        case AssetCalibrationForecastKey.Day:
          field.values = field.values!.map(this.formatDateForDay)
          break;
        case AssetCalibrationForecastKey.Week:
          field.values = field.values!.map(this.formatDateForWeek)
          break;
        case AssetCalibrationForecastKey.Month:
          field.values = field.values!.map(this.formatDateForMonth)
          break;
        default:
          break;
      }
    });
  }

  processResultsGroupedByProperties(result: DataFrameDTO) {
    const formattedFields = [] as FieldDTO[];
    formattedFields.push({ name: "Group", values: [] } as FieldDTO);
    formattedFields.push({ name: "Assets", values: [] } as FieldDTO);

    for (let columnIndex = 0; columnIndex < result.fields.length; columnIndex++) {
      const columnName = this.createColumnNameFromDescriptor(result.fields[columnIndex] as FieldDTOWithDescriptor);
      const columnValue = result.fields[columnIndex].values?.at(0);
      formattedFields[0].values!.push(columnName);
      formattedFields[1].values!.push(columnValue);
    }

    result.fields = formattedFields;
  }

  createColumnNameFromDescriptor(field: FieldDTOWithDescriptor): string {
    return field.columnDescriptors.map(descriptor => {
      if (descriptor.type === ColumnDescriptorType.MinionId && this.systemAliasCache) {
        const system = this.systemAliasCache.get(descriptor.value);

        return system?.alias || descriptor.value
      }
      return descriptor.value
    }).join(' - ');
  }

  formatDateForDay(date: string): string {
    return new Date(date).toISOString().split('T')[0];
  }

  formatDateForWeek(date: string): string {
    const startDate = new Date(date);
    const endDate = new Date(startDate);
    endDate.setDate(startDate.getDate() + 6);
    return `${startDate.toISOString().split('T')[0]} : ${endDate.toISOString().split('T')[0]}`;
  }

  formatDateForMonth(date: string): string {
    return new Date(date).toLocaleDateString('en-US', { year: 'numeric', month: 'long' });
  }

  shouldRunQuery(calibrationQuery: AssetCalibrationQuery): boolean {
    return calibrationQuery.groupBy.length > 0;
  }

  async queryAssets(filter = '', take = -1): Promise<AssetModel[]> {
    let data = { filter, take };
    try {
      let response = await this.post<AssetsResponse>(this.baseUrl + '/query-assets', data);
      return response.assets;
    } catch (error) {
      throw new Error(`An error occurred while querying assets: ${error}`);
    }
  }

  async queryCalibrationForecast(groupBy: string[], startTime: string, endTime: string, filter = ''): Promise<CalibrationForecastResponse> {
    let data = { groupBy, startTime, endTime, filter };
    try {
      let response = await this.post<CalibrationForecastResponse>(this.baseUrl + '/assets/calibration-forecast', data);
      return response;
    } catch (error) {
      throw new Error(`An error occurred while querying assets calibration forecast: ${error}`);
    }
  }

  async querySystems(filter = '', projection = defaultProjection): Promise<SystemMetadata[]> {
    try {
      let response = await this.getSystems({
        filter: filter,
        projection: `new(${projection.join()})`,
        orderBy: defaultOrderBy,
      })

      return response.data;
    } catch (error) {
      throw new Error(`An error occurred while querying systems: ${error}`);
    }
  }

  private async loadSystems(): Promise<void> {
    if (this.systemAliasCache.size > 0) {
      return;
    }

    const systems = await this.querySystems('', ['id', 'alias','connected.data.state', 'workspace']);
    systems.forEach(system => this.systemAliasCache.set(system.id, system));
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    await this.get(this.baseUrl + '/assets?take=1');
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }
}
