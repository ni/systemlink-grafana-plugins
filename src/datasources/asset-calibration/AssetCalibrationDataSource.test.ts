import { BackendSrv } from "@grafana/runtime";
import { MockProxy } from "jest-mock-extended";
import {
  createFetchError,
  createFetchResponse,
  getQueryBuilder,
  requestMatching,
  setupDataSource,
} from "test/fixtures";
import { AssetCalibrationDataSource } from "./AssetCalibrationDataSource";
import {
  AssetCalibrationPropertyGroupByType,
  AssetCalibrationQuery,
  AssetCalibrationTimeBasedGroupByType,
  CalibrationForecastResponse,
} from "./types";

let datastore: AssetCalibrationDataSource, backendServer: MockProxy<BackendSrv>

beforeEach(() => {
  [datastore, backendServer] = setupDataSource(AssetCalibrationDataSource);
});

const monthGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
            { name: "Month", values: ["2022-01-01T00:00:00.0000000Z", "2022-02-01T00:00:00.0000000Z", "2022-03-01T00:00:00.0000000Z"] },
            { name: "Assets", values: [1, 0, 3] },
        ]
    }
}

const weekGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
            { name: "Day", values: ["2022-01-01T00:00:00.0000000Z", "2022-01-02T00:00:00.0000000Z", "2022-01-03T00:00:00.0000000Z"] },
            { name: "Assets", values: [1, 2, 2] },
        ]
    }
}

const dayGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
            { name: "Week", values: ["2022-01-03T00:00:00.0000000Z", "2022-01-10T00:00:00.0000000Z", "2022-01-17T00:00:00.0000000Z"] },
            { name: "Assets", values: [1, 2, 2] },
        ]
    }
}

const locationGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
          { name: "Location1", values: [1] },
          { name: "Location2", values: [2] },
          { name: "Location3", values: [3] },
        ]
    }
}

const modelGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
          { name: "Model1", values: [1] },
          { name: "Model2", values: [2] },
        ]
    }
}

const emptyGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
        ]
    }
}

const modelLocationGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
          { name: "Model1 - Localtion1", values: [1] },
          { name: "Model2 - Localtion1", values: [2] },
        ]
    }
}

const monthLocationGroupCalibrationForecastResponseMock: CalibrationForecastResponse =
{
    calibrationForecast: {
        columns: [
            { name: "Month", values: ["2022-01-01T00:00:00.0000000Z", "2022-02-01T00:00:00.0000000Z", "2022-03-01T00:00:00.0000000Z"] },
            { name: "Location1", values: [1, 2, 3] },
            { name: "Location2", values: [2, 4, 1] },
            { name: "Location3", values: [4, 3, 1] },
        ]
    }
}

const monthBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationTimeBasedGroupByType.Month],
}

const weekBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationTimeBasedGroupByType.Week],
}

const dayBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationTimeBasedGroupByType.Day],
}

const locationBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationPropertyGroupByType.Location],
}

const modelBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationPropertyGroupByType.Model],
}

const modelLocationBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationPropertyGroupByType.Model, AssetCalibrationPropertyGroupByType.Location],
}

const monthLocationBasedCalibrationForecastQueryMock: AssetCalibrationQuery = {
  refId: '',
  groupBy: [AssetCalibrationTimeBasedGroupByType.Month, AssetCalibrationPropertyGroupByType.Location],
}

const buildCalibrationForecastQuery = getQueryBuilder<AssetCalibrationQuery>()({
  refId: '',
  groupBy: []
});

describe('testDatasource', () => {
  test('returns success', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets?take=1' }))
      .mockReturnValue(createFetchResponse(25));

    const result = await datastore.testDatasource();

    expect(result.status).toEqual('success');
  });

  test('bubbles up exception', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets?take=1' }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.testDatasource()).rejects.toHaveProperty('status', 400);
  });
})

describe('queries', () => {
  test('asset calibration forecast with month groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(monthGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(monthBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('asset calibration forecast with week groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(weekGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(weekBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('asset calibration forecast with day groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(dayGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(dayBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('calibration forecast with location groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(locationGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(locationBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('calibration forecast with model groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(modelGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(modelBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('calibration forecast with model and location groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(modelLocationGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(modelLocationBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('calibration forecast with month and location groupBy', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(monthLocationGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(monthLocationBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('calibration forecast with month groupBy returns empty results', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchResponse(emptyGroupCalibrationForecastResponseMock as CalibrationForecastResponse))

    const result = await datastore.query(buildCalibrationForecastQuery(monthBasedCalibrationForecastQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('handles metadata query error', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets/calibration-forecast' }))
      .mockReturnValue(createFetchError(418))

    await expect(datastore.query(buildCalibrationForecastQuery(monthBasedCalibrationForecastQueryMock))).rejects.toThrow()
  })
})
