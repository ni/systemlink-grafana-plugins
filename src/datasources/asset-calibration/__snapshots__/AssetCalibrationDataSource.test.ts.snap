// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries asset calibration forecast with day groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Week",
        "values": [
          "2022-01-03 : 2022-01-09",
          "2022-01-10 : 2022-01-16",
          "2022-01-17 : 2022-01-23",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          2,
          2,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries asset calibration forecast with month groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Month",
        "values": [
          "January 2022",
          "February 2022",
          "March 2022",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          0,
          3,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries asset calibration forecast with week groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Day",
        "values": [
          "2022-01-01",
          "2022-01-02",
          "2022-01-03",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          2,
          2,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries calibration forecast with location groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Group",
        "values": [
          "Location1",
          "Location2",
          "Location3",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          2,
          3,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries calibration forecast with model and location groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Group",
        "values": [
          "Model1 - Localtion1",
          "Model2 - Localtion1",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          2,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries calibration forecast with model groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Group",
        "values": [
          "Model1",
          "Model2",
        ],
      },
      {
        "name": "Assets",
        "values": [
          1,
          2,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries calibration forecast with month and location groupBy 1`] = `
[
  {
    "fields": [
      {
        "name": "Month",
        "values": [
          "January 2022",
          "February 2022",
          "March 2022",
        ],
      },
      {
        "name": "Location1",
        "values": [
          1,
          2,
          3,
        ],
      },
      {
        "name": "Location2",
        "values": [
          2,
          4,
          1,
        ],
      },
      {
        "name": "Location3",
        "values": [
          4,
          3,
          1,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries calibration forecast with month groupBy returns empty results 1`] = `
[
  {
    "fields": [],
    "refId": "A",
  },
]
`;
